pkg/dns/dns.go

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dns
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/miekg/dns"
+
+	logger "istio.io/istio/pkg/log"
+)
+
+var log = logger.RegisterScope("kmesh-dns", "kmesh-dns dns")
+
+type KmeshDNSServer struct {
+	dnsProxies []*dnsProxy
+	mapMutex   sync.RWMutex
+	dnsEntries map[string][]net.IP
+}
+
+func NewDNSServer(addr string) (*KmeshDNSServer, error) {
+	h := &KmeshDNSServer{}
+	if addr == "" {
+		addr = "localhost:15053"
+	}
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, fmt.Errorf("dns address must be a valid host:port: %v", err)
+	}
+	addresses := []string{addr}
+	if host == "localhost" {
+		addresses = []string{}
+		addresses = append(addresses, net.JoinHostPort("0.0.0.0", port))
```

Comment:
? should you remove local host? If address == "" or address == ":15053", or `0.0.0:15053`, make them listen on 0.0.0:15053